src/collapse/perf.rs

```diff
@@ -1,114 +1,171 @@
-#[derive(Default)]
 pub struct Folder {
-    /// All lines until the next empty line are stack lines.
-    in_event: bool,
-
-    /// Skip all stack lines in this event.
-    skip_stack: bool,
+    // State...
+    /// General String cache that can be used while processing lines. Currently only used to keep
+    /// track of functions for Java inlining.
+    cache_line: Vec<String>,
 
-    /// Function entries on the stack in this entry thus far.
-    stack: VecDeque<String>,
+    /// Similar to, but different from, the `event_filter` field on `Options`
+    ///
+    /// * Field on `Options` represents user's provided configuration and will never change.
+    /// * This field, on the other hand, although initially set to the user's provided
+    ///   configuration, represents state and may change as data is processed. In particular, it can
+    ///   change from `None` initially to `Some(...)` as data is processed.
+    event_filter: Option<String>,
 
-    /// General String cache that can be used while processing lines.
-    /// Currently only used to keep track of functions for Java inlining.
-    cache_line: Vec<String>,
+    /// All lines until the next empty line are stack lines.
+    in_event: bool,
 
     /// Number of times each call stack has been seen.
-    occurrences: FnvHashMap<String, usize>,
+    occurrences: Occurrences,
 
     /// Current comm name.
     ///
     /// Called pname after original stackcollapse-perf source.
     pname: String,
 
-    event_filtering: EventFilterState,
+    /// Skip all stack lines in this event.
+    skip_stack: bool,
+
+    /// Function entries on the stack in this entry thus far.
+    stack: VecDeque<String>,
 
+    // Options...
     opt: Options,
 }
 
+impl From<Options> for Folder {
+    fn from(mut opt: Options) -> Self {
+        if opt.nstacks_per_job == 0 {
+            opt.nstacks_per_job = 1;
+        }
+        if opt.nthreads == 0 {
+            opt.nthreads = 1;
+        }
+        let occurrences = if opt.nthreads < 2 {
+            Occurrences::new_single_threaded()
+        } else {
+            Occurrences::new_multi_threaded()
+        };
+        opt.include_pid = opt.include_pid || opt.include_tid;
+        Self {
+            cache_line: Vec::default(),
+            event_filter: opt.event_filter.clone(),
+            in_event: false,
+            occurrences,
+            pname: String::default(),
+            skip_stack: false,
+            stack: VecDeque::default(),
+            opt,
+        }
+    }
+}
+
+impl Default for Folder {
+    fn default() -> Self {
+        Options::default().into()
+    }
+}
+
 impl Collapse for Folder {
-    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    fn collapse<R, W>(&mut self, reader: R, writer: W) -> io::Result<()>
     where
         R: BufRead,
         W: Write,
     {
-        let mut line = String::new();
-        loop {
-            line.clear();
+        if self.opt.nthreads < 2 {
```

Comment:
This should use `self.occurrences.is_concurrent` rather than writing out the same conditional again.